### DISCUSSION TITLE
onButtonPress and onButtonRelease (also: better bike horn honking)

### DISCUSSION THREAD

**Discussion by aoyamajia** [FakeFishGames/Barotrauma#14249] [Ideas and suggestions]
tl;dr It would be more straightforward to make items with similar functionality if we could more easily detect when LeftMouse has been pressed or released.

---
I think there is modding potential in allowing status effects to trigger on button press and on button release. Examples that can use this are the bike horn, Startup and Cooldown sounds in a hypothetical handheld chaingun (RangedWeapon), a hypothetical `<StatusEffect type="onButtonPress" button="MiddleMouse" />` to fire an underbarrel grenade launcher, or in general any sort of charged attacks (click to charge, release to fire).

I have a [mod](https://steamcommunity.com/sharedfiles/filedetails/?id=3280888044) that sort of emulates this behaviour, allowing the bike horn to behave more like how a real-world bike horn behaves:
```xml
<!-- high reload time to prevent honk_1 from playing again until you release LeftMouse -->
<RangedWeapon reload="1000">
  <!-- `honked` tag is for detecting that we have clicked,
       it will be removed when LeftMouse is released -->
  <StatusEffect tags="honked" type="onUse" target="This">
    <Sound file="%ModDir%/Honk/Sounds/honk_1.ogg" range="1000" />
  </StatusEffect>
  <StatusEffect tags="honking" duration="0.2" type="onUse" target="This" />
</RangedWeapon>

<!-- detect when LeftMouse is being held. -->
<!-- Has to be done in a LightComponent or similar component,
     as onUse in RangedWeapon is only triggered once per cooldown,
     whereas here it is triggered constantly -->
<LightComponent>
  <StatusEffect tags="honk_primed" duration="0.05" type="onUse" target="This" />
  <!-- you could theoretically also put the regular tag `honked` in its own statuseffect here. -->
</LightComponent>

<!-- detect when LeftMouse has been released. -->
<ItemComponent>
  <IsActive hasStatusTag="neq honk_primed" />
  <IsActive hasTag="honked" />
  <StatusEffect tags="" type="onActive" target="This" disableDeltaTime="True"
                setValue="True" reloadTimer="0.15" targetItemComponent="RangedWeapon">
    <!-- additionally check if not currently playing honk_1, so that the honks do not play simultaneously -->
    <Conditional hasStatusTag="neq honking" />
    <Sound file="%ModDir%/Honk/Sounds/honk_2.ogg" range="1000" />
  </StatusEffect>
</ItemComponent>
```
resulting in the following (22s 1.17MB mp4):

https://github.com/FakeFishGames/Barotrauma/assets/39614559/1d837160-4ba5-4ac5-9a0f-7e2e861a7ca2

**Comment by aoyamajia** [comment]
While attempting to write this myself, I just realised that an *actual* onButtonPress/onButtonRelease would only be useable by player characters; so perhaps instead of those, it would be more apt to have:

- onUseStart: when an item starts being used (i.e., onButtonPress for players, vanilla onUse for mostly everything else)
- onUse: in this case, when a player is holding left click or when a bot is using a constant item without cooldown, like a repair tool
- onUseEnd: when an item stops being used (i.e., onButtonRelease for players, or when a bot stops using a repair tool)

I understand that this would definitely be quite a lot of effort just to support niche/modding use cases, and little to no improvements for vanilla items aside from potentially having to edit them all to use onUseStart if the effects are supposed to be one-time per "use."
Additionally, there are other issues I can vaguely think of regarding backwards compatibility with vanilla onUse, which I cannot put to words at the moment.

**Comment by zhu-rengong** [comment]
I have a way to make ho+nk easier:
```xml
<Holdable slots="Any,LeftHand,RightHand" holdangle="90" aimpos="60,-15" handle1="-20,0" msg="ItemMsgPickUpSelect">
  <StatusEffect type="OnUse" target="This">
    <Sound file="Content/Items/Button/Switch1.ogg" type="OnUse" range="1050" frequency="1.45" volume="0.8" />
    <Conditional hasstatustag="! e1" />
  </StatusEffect>
  <StatusEffect type="OnUse" target="This" tags="e1" duration=".033" stackable="false" />
  <StatusEffect type="OnActive" target="This" comparison="And">
    <Sound file="Content/Items/Button/Switch2.ogg" type="OnUse" range="1050" frequency="1.45" volume="0.8" />
    <Conditional hasstatustag="! e1" />
    <Conditional hasstatustag="e2" />
  </StatusEffect>
  <StatusEffect type="OnActive" target="This" tags="e2" duration=".033" stackable="false">
    <Conditional hasstatustag="e1" />
  </StatusEffect>
</Holdable>
```
No need to consider too much compatibility, but it still took a long way to achieve it.

**Reply by aoyamajia** [reply]
I'm not sure why the vanilla bike horn uses a RangedWeapon (but I think it's hilarious that it does), so I kept it in and only added to it. If there's no particular reason, then this is a better implementation, yes.
